sandbox/cgroup/tracking_test.go

```diff
@@ -0,0 +1,580 @@
+	dirs.SetRootDir("")
+}
+
+// CreateTransientScopeForTracking is a no-op when refresh app awareness is off
+func (s *trackingSuite) TestCreateTransientScopeForTrackingFeatureDisabled(c *C) {
+	noDBus := func() (*dbus.Conn, error) {
+		return nil, fmt.Errorf("dbus should not have been used")
+	}
+	restore := dbusutil.MockConnections(noDBus, noDBus)
+	defer restore()
+
+	c.Assert(features.RefreshAppAwareness.IsEnabled(), Equals, false)
+	err := cgroup.CreateTransientScopeForTracking("snap.pkg.app")
+	c.Check(err, IsNil)
+}
+
+// CreateTransientScopeForTracking does stuff when refresh app awareness is on
+func (s *trackingSuite) TestCreateTransientScopeForTrackingFeatureEnabled(c *C) {
+	// Pretend that refresh app awareness is enabled
+	enableFeatures(c, features.RefreshAppAwareness)
+	c.Assert(features.RefreshAppAwareness.IsEnabled(), Equals, true)
+	// Pretend we are a non-root user so that session bus is used.
+	restore := cgroup.MockOsGetuid(12345)
+	defer restore()
+	// Pretend our PID is this value.
+	restore = cgroup.MockOsGetpid(312123)
+	defer restore()
+	// Rig the random UUID generator to return this value.
+	uuid := "cc98cd01-6a25-46bd-b71b-82069b71b770"
+	restore = cgroup.MockRandomUUID(uuid)
+	defer restore()
+	// Replace interactions with DBus so that only session bus is available and responds with our logic.
+	conn, err := dbustest.Connection(func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
+			return []*dbus.Message{checkAndRespondToStartTransientUnit(c, msg, "snap.pkg.app."+uuid+".scope", 312123)}, nil
+		}
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	c.Assert(err, IsNil)
+	restore = dbusutil.MockOnlySessionBusAvailable(conn)
+	defer restore()
+	// Replace the cgroup analyzer function
+	restore = cgroup.MockCgroupProcessPathInTrackingCgroup(func(pid int) (string, error) {
+		return "/user.slice/user-12345.slice/user@12345.service/snap.pkg.app." + uuid + ".scope", nil
+	})
+	defer restore()
+
+	err = cgroup.CreateTransientScopeForTracking("snap.pkg.app")
+	c.Check(err, IsNil)
+}
+
+func (s *trackingSuite) TestCreateTransientScopeForTrackingUnhappyNotRootGeneric(c *C) {
+	// Pretend that refresh app awareness is enabled
+	enableFeatures(c, features.RefreshAppAwareness)
+
+	// Hand out stub connections to both the system and session bus.
+	// Neither is really used here but they must appear to be available.
+	restore := dbusutil.MockConnections(dbustest.StubConnection, dbustest.StubConnection)
+	defer restore()
+
+	// Pretend we are a non-root user so that session bus is used.
+	restore = cgroup.MockOsGetuid(12345)
+	defer restore()
+	// Pretend our PID is this value.
+	restore = cgroup.MockOsGetpid(312123)
+	defer restore()
+
+	// Disable the cgroup analyzer function as we don't expect it to be used in this test.
+	restore = cgroup.MockCgroupProcessPathInTrackingCgroup(func(pid int) (string, error) {
+		panic("we are not expecting this call")
```

Comment:
a nitpick, but I think when tests fail with this sort of thing it's more readable to see `c.Error("test should not have done this")` and `return "", fmt.Errorf("test broken")` as that will give the reader more readable information from gocheck about how it failed, otherwise you just see a bunch of panic stack traces, etc. then the failure info from gocheck.

Reply:
Indeed, I adjusted all the panics in this file.